src/derive_utils.rs

```diff
@@ -34,76 +34,88 @@ pub struct ParamDescription {
-            if fname.is_some() { "()" } else { "" },
-            params.len(),
-            if params.len() == 1 { "s" } else { "" },
-            nargs + nkeywords
-        )));
-    }
-    let mut used_keywords = 0;
+    let mut used_args = 0;
     // Iterate through the parameters and assign values to output:
     for (i, (p, out)) in params.iter().zip(output).enumerate() {
         match kwargs.and_then(|d| d.get_item(p.name)) {
             Some(kwarg) => {
                 *out = Some(kwarg);
-                used_keywords += 1;
                 if i < nargs {
                     return Err(TypeError::py_err(format!(
-                        "Argument given by name ('{}') and position ({})",
-                        p.name,
-                        i + 1
+                        "{} got multiple values for argument '{}'",
+                        fname.unwrap_or("function"),
+                        p.name
                     )));
                 }
+                kwargs.as_ref().unwrap().del_item(p.name).unwrap();
             }
             None => {
                 if p.kw_only {
                     if !p.is_optional {
                         return Err(TypeError::py_err(format!(
-                            "Required argument ('{}') is keyword only argument",
+                            "{} missing required keyword-only argument '{}'",
+                            fname.unwrap_or("function"),
                             p.name
                         )));
                     }
                     *out = None;
                 } else if i < nargs {
                     *out = Some(args.get_item(i));
+                    used_args += 1;
                 } else {
                     *out = None;
                     if !p.is_optional {
                         return Err(TypeError::py_err(format!(
-                            "Required argument ('{}') (pos {}) not found",
-                            p.name,
-                            i + 1
+                            "{} missing required positional argument: '{}'",
+                            fname.unwrap_or("function"),
+                            p.name
                         )));
                     }
                 }
             }
         }
     }
-    if !accept_kwargs && used_keywords != nkeywords {
-        // check for extraneous keyword arguments
-        for item in kwargs.unwrap().items().iter() {
-            let item = <PyTuple as PyTryFrom>::try_from(item)?;
-            let key = <PyString as PyTryFrom>::try_from(item.get_item(0))?.to_string()?;
-            if !params.iter().any(|p| p.name == key) {
-                return Err(TypeError::py_err(format!(
-                    "'{}' is an invalid keyword argument for this function",
-                    key
-                )));
-            }
+    // check for extraneous keyword arguments
+    if !accept_kwargs && !kwargs.as_ref().map_or(true, |d| d.is_empty()) {
```

Comment:
Since `d.is_empty()` is called later again, how about defining `let kwargs_is_empty = kwargs.as_ref().map_or(false, |d| d.is_empty());`?